File: app/src/main/java/com/bear2b/sampleapp/ui/view/fragments/SuccessScanFragment.java
package com.bear2b.sampleapp.ui.view.fragments;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Fragment;
import android.content.DialogInterface;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ShareCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import com.bear.common.sdk.BearHandler;
import com.bear2b.sampleapp.R;
import com.bear2b.sampleapp.ui.view.activities.AdvancedSampleActivity;

import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.OnNeverAskAgain;
import permissions.dispatcher.OnPermissionDenied;
import permissions.dispatcher.OnShowRationale;
import permissions.dispatcher.PermissionRequest;
import permissions.dispatcher.RuntimePermissions;

@RuntimePermissions
public class SuccessScanFragment extends Fragment {

    private BearHandler handler;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        handler = ((AdvancedSampleActivity) getActivity()).getHandler();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.success_scan_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Button btnScreenshot = (Button) view.findViewById(R.id.btnScreenshot);
        btnScreenshot.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SuccessScanFragmentPermissionsDispatcher.shareImageWithCheck(SuccessScanFragment.this);
            }
        });
        Button btnSamplePausingFragment = (Button) view.findViewById(R.id.btnSamplePausingFragment);
        btnSamplePausingFragment.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ((AdvancedSampleActivity) getActivity()).pauseArView();
                ((AdvancedSampleActivity) getActivity()).showSamplePausingFragment();
            }
        });
    }

    @Override
    public void onPause() {
        // Screenshot may not be completed when fragment is paused
        handler.cancelScreenshot();
        super.onPause();
    }

    public static SuccessScanFragment newInstance() {
        return new SuccessScanFragment();
    }

    @NeedsPermission({"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"})
    public void shareImage() {
        handler.takeScreenshot(new BearHandler.ScreenshotCallback() {
            @Override
            public void onComplete(@NonNull Uri imageUri) {
                ShareCompat.IntentBuilder
                        .from(getActivity())
                        .setType("image/*")
                        .setStream(imageUri)
                        .startChooser();
            }
        });
    }

    @OnShowRationale({"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"})
    void showRationaleForCamera(final PermissionRequest request) {
        new AlertDialog.Builder(getActivity())
                .setMessage("request permission")
                .setPositiveButton("allow", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.proceed();
                    }
                })
                .setNegativeButton("deny", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.cancel();
                    }
                })
                .show();
    }

    @OnPermissionDenied({"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"})
    void showDeniedStorage() {
        Toast.makeText(getActivity(), "permission denied", Toast.LENGTH_SHORT).show();
    }

    @OnNeverAskAgain({"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"})
    void showNeverAskForStorage() {
        Toast.makeText(getActivity(), "never ask again", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        SuccessScanFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }
}

File: app/src/kotlin/java/com/bear2b/sampleapp/ui/view/fragments/SuccessScanFragment.kt
package com.bear2b.sampleapp.ui.view.fragments

import android.app.AlertDialog
import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.Nullable
import androidx.core.app.ShareCompat
import androidx.fragment.app.Fragment
import com.bear.common.sdk.IBearHandler
import com.bear2b.sampleapp.R
import com.bear2b.sampleapp.ui.view.activities.AdvancedSampleActivity
import kotlinx.android.synthetic.main.success_scan_fragment.*
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.OnNeverAskAgain
import permissions.dispatcher.OnPermissionDenied
import permissions.dispatcher.OnShowRationale
import permissions.dispatcher.PermissionRequest
import permissions.dispatcher.RuntimePermissions

@RuntimePermissions
class SuccessScanFragment : Fragment() {

    private lateinit var bearHandler : IBearHandler

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        bearHandler = (activity as AdvancedSampleActivity).bearHandler
    }

    @Nullable
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.success_scan_fragment, container,false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        btnScreenshot.setOnClickListener { SuccessScanFragmentPermissionsDispatcher.shareImageWithCheck(this@SuccessScanFragment) }
        btnSamplePausingFragment.setOnClickListener {
            (activity as AdvancedSampleActivity).pauseArView()
            (activity as AdvancedSampleActivity).showSamplePausingFragment()
        }
    }

    override fun onPause() {
        // Screenshot may not be completed when fragment is paused
        bearHandler.cancelScreenshot()
        super.onPause()
    }

    companion object {
        fun newInstance() = SuccessScanFragment()
    }

    @NeedsPermission("android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE")
    fun shareImage() = bearHandler.takeScreenshot(object : IBearHandler.ScreenshotCallback {
        override fun onComplete(imageUri: Uri) {
            ShareCompat.IntentBuilder
                    .from(activity)
                    .setType("image/*")
                    .setStream(imageUri)
                    .startChooser()
        }

        override fun onCancel() {
            Log.e("DEV", "Screenshot has been cancelled")
        }
    })

    @OnShowRationale("android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE")
    fun showRationaleForCamera(request: PermissionRequest) {
        AlertDialog.Builder(activity)
                .setMessage("request permission")
                .setPositiveButton("allow") { _, _ -> request.proceed() }
                .setNegativeButton("deny") { _, _ -> request.cancel() }
                .show()
    }

    @OnPermissionDenied("android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE")
    fun showDeniedStorage() = Toast.makeText(activity, "permission denied", Toast.LENGTH_SHORT).show()

    @OnNeverAskAgain("android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE")
    fun showNeverAskForStorage() = Toast.makeText(activity, "never ask again", Toast.LENGTH_SHORT).show()

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        SuccessScanFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults)
    }
}